akka-stream/src/main/scala/akka/stream/impl/FanIn.scala

```diff
@@ -46,22 +53,39 @@ private[akka] object FanIn {
       }
     }
 
-    private val marked = Array.ofDim[Boolean](inputCount)
+    private[this] final val states = Array.ofDim[State](inputCount)
     private var markCount = 0
-    private val pending = Array.ofDim[Boolean](inputCount)
     private var markedPending = 0
-    private val depleted = Array.ofDim[Boolean](inputCount)
-    private val completed = Array.ofDim[Boolean](inputCount)
     private var markedDepleted = 0
-    private val cancelled = Array.ofDim[Boolean](inputCount)
+
+    private[this] final def hasState(index: Int, flag: Int): Boolean =
+      (states(index) & flag) != 0
+    private[this] final def setState(index: Int, flag: Int, on: Boolean): Unit =
+      states(index) = if (on) (states(index) | flag).toByte else (states(index) & ~flag).toByte
```

Comment:
@rkuhn the downside of your proposal is that every write needs to be preceded by a read.